File: 3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = set()
        l, r = 0, 0
        ans = 0
        while r < len(s):
            while s[r] in seen:
                seen.remove(s[l])
                l += 1
            ans = max(ans, r - l + 1)
            seen.add(s[r])
            r += 1
        return ans

File: 3.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    vector<int> v(128);
    int l = 0, r = 0;
    int ans = 0;
    while (r < s.size()) {
      ++v[s[r]];
      while (v[s[r]] > 1) {
        --v[s[l++]];
      }
      ans = max(ans, r - l + 1);
      ++r;
    }
    return ans;
  }
};

int main() {
  cout << "test 1, expect 3" << endl;
  cout << Solution().lengthOfLongestSubstring("abcabcbb") << endl;
}